.github/workflows/build-and-test.yml

```diff
@@ -34,6 +34,35 @@ jobs:
     - name: Build all classes
       run: ./gradlew --stacktrace classes
 
+    - name: Install Node
+      uses: actions/setup-node@v4
+      with:
+        node-version: 20
+
+    - uses: pnpm/action-setup@v3
+      name: Install pnpm
+      with:
+        version: 8
+        run_install: false
+
+    - name: Get pnpm store directory
+      shell: bash
+      run: |
+        echo "STORE_PATH=$(pnpm store path --silent)" >> $GITHUB_ENV
+
+    - uses: actions/cache@v4
+      name: Setup pnpm cache
+      with:
+        path: ${{ env.STORE_PATH }}
+        key: ${{ runner.os }}-pnpm-store-${{ hashFiles('**/pnpm-lock.yaml') }}
+        restore-keys: |
+          ${{ runner.os }}-pnpm-store-
+
+    - name: Install UI dependencies
+      run: |
+        cd ui
+        pnpm install
```

Comment:
Maybe making use of the [`-C` / `--dir` option](https://pnpm.io/pnpm-cli#-c-path---dir-path) would make up for a slight more elegant one-liner.